generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java

```diff
@@ -0,0 +1,551 @@
+                    continue;
+                generateResponseTest(operation, httpResponseTestCase);
+            }
+        });
+    }
+
+    private void generateResponseTest(OperationShape operation, HttpResponseTestCase httpResponseTestCase) {
+        if (httpResponseTestCase.getDocumentation().isPresent()) {
+            writer.writeXmlDocs(httpResponseTestCase.getDocumentation().get());
+        }
+        if (ProtocolTestCustomizations.VNextTests.contains(httpResponseTestCase.getId())) {
+            writer.writeSingleLineComment("This test requires a breaking change, and will be addressed in V4");
+            writer.write("[Ignore]");
+        }
+        writer.write("[TestMethod]");
+        writer.write("[TestCategory(\"ProtocolTest\")]");
+        writer.write("[TestCategory(\"ResponseTest\")]");
+        writer.write("[TestCategory(\"$L\")]", serviceName);
+        writer.openBlock("public void $LResponse()\n{", "}", httpResponseTestCase.getId(), () -> {
+            generateResponseTestBlock(operation, httpResponseTestCase);
+        });
+        writer.write("\n");
+    }
+
+    private void generateResponseTestBlock(OperationShape operation, HttpResponseTestCase httpResponseTestCase) {
+        var outputShape = model.expectShape(operation.getOutputShape(), StructureShape.class);
+        var responseSymbol = operation.getId().getName() + "Response";
+        //Arrange
+        arrangeResponseTestBlock(httpResponseTestCase);
+        writer.write("\n");
+        // only unmarshall the response and assert if a body is present, as per smithy spec
+        if (httpResponseTestCase.getBody().isPresent() && !httpResponseTestCase.getBody().equals("")) {
+            //Act
+            writer.writeSingleLineComment("Act");
+            writer.write("var unmarshalledResponse = new $LUnmarshaller().Unmarshall(context);", responseSymbol);
+            writer.openBlock("var expectedResponse = new $L\n{", "};", responseSymbol, (Runnable) () -> httpResponseTestCase.getParams().accept(new ValueNodeVisitor(outputShape, true, responseSymbol)));
+            writer.write("\n");
+            //Assert
+            writer.writeSingleLineComment("Assert");
+            writer.write("var actualResponse = ($L)unmarshalledResponse;", responseSymbol);
+            writer.write("Comparer.CompareObjects<$L>(expectedResponse,actualResponse);", responseSymbol);
+        }
+        if (!httpResponseTestCase.getForbidHeaders().isEmpty()) {
+            for (var forbidHeader : httpResponseTestCase.getForbidHeaders()) {
+                writer.write("Assert.IsFalse(context.ResponseData.IsHeaderPresent($S);", forbidHeader);
+            }
+        }
+        if (!httpResponseTestCase.getRequireHeaders().isEmpty()) {
+            for (var requireHeader : httpResponseTestCase.getRequireHeaders()) {
+                writer.write("Assert.IsTrue(context.ResponseData.IsHeaderPresent($S);", requireHeader);
+            }
+        }
```

Comment:
sorry, looking at this code I realized I did not think through this fully. I saw the httpResponseTests can have required/forbid headers and didn't see support for it in the generator, so pointed it out. But now I realize that you are only generating client tests. required/forbid headers for httpResponseTests would be something to assert on in server tests. They are not relevant to client tests. So please remove this. Sorry for that.